backend/python/tapa/codegen/duplicate_s_axi_control.py

```diff
@@ -0,0 +1,317 @@
+from typing import Iterator, List
+
+from tapa.task import Task
+from pyverilog.vparser.ast import (
+  InstanceList,
+  PortArg,
+  Node,
+  ModuleDef,
+  Decl,
+  Wire,
+  Input,
+  Output,
+  Parameter,
+  Instance,
+  Identifier,
+  ParamArg,
+)
+
+DONT_TOUCH_PORTS = (
+  'CLK',
+  'ACLK',
+  'ARESET',
+  'RESET',
+  'ACLK_EN',
+  'ap_clk',
+  'ap_rst',
+  'ap_rst_n'
+)
+
+S_AXI_AW_AND_W_PORTS = (
+  'AWVALID',
+  'AWREADY',
+  'AWADDR',
+  'WVALID',
+  'WREADY',
+  'WDATA',
+  'WSTRB',
+)
+
+S_AXI_AR_AND_R_PORTS = (
+  'ARVALID',
+  'ARREADY',
+  'ARADDR',
+  'RVALID',
+  'RREADY',
+  'RDATA',
+  'RRESP',
+)
+
+S_AXI_B_PORTS = (
+  'BVALID',
+  'BREADY',
+  'BRESP',
+)
+
+CONTROL_SIGNALS = (
+  'ap_start',
+  'interrupt',
+  'ap_ready',
+  'ap_done',
+  'ap_idle',
+)
+
+
+def _get_nodes(node, *target_types) -> Iterator[Node]:
+  for target_type in target_types:
+    if isinstance(node, target_type):
+      yield node
+  for c in node.children():
+    yield from _get_nodes(c, *target_types)
+
+
+def get_ctrl_instance(top_task: Task, ctrl_inst_name) -> InstanceList:
+  ctrl_instance_list = []
+  for inst_list in top_task.module.get_nodes_of_type(InstanceList):
+    if inst_list.module == ctrl_inst_name:
+      ctrl_instance_list.append(inst_list)
+
+  if len(ctrl_instance_list) != 1:
+    raise NotImplementedError('%d control_s_axi instances detected', len(ctrl_instance_list))
+
+  return ctrl_instance_list[0]
+
+
+def get_updated_ctrl_instance(
+    ctrl_instance: InstanceList,
+    suffix: str,
+    is_root: bool,
+) -> InstanceList:
+  """ append a suffix to the argname of some PortArgs """
+  inst_list = deepcopy(ctrl_instance)
+
+  # update instance name. inst.name is not an Identifier node but a str
+  assert len(inst_list.instances) == 1
+  inst_list.instances[0].name += suffix
+
+  if is_root:
+    for port_arg in _get_nodes(inst_list, PortArg):
+      if any(re.match(p, port_arg.portname) for p in
+          (DONT_TOUCH_PORTS +
+          CONTROL_SIGNALS +
+          S_AXI_AR_AND_R_PORTS +
+          S_AXI_B_PORTS)
+      ):
+        continue
+      else:
+        port_arg.argname.name += suffix
+
+  else:
+    for port_arg in _get_nodes(inst_list, PortArg):
+      if any(re.match(p, port_arg.portname) for p in DONT_TOUCH_PORTS):
+        continue
+      elif port_arg.portname in S_AXI_AR_AND_R_PORTS:
+        if port_arg.portname == 'ARVALID':
+          port_arg.argname.name = '1\'b0'
+        else:
+          port_arg.argname.name = ''
+      elif port_arg.portname in S_AXI_B_PORTS:
+        if port_arg.portname == 'BREADY':
+          port_arg.argname.name = '1\'b1'
+        else:
+          port_arg.argname.name = ''
+      elif port_arg.portname in CONTROL_SIGNALS:
+        port_arg.argname.name = ''
+      else:
+        port_arg.argname.name += suffix
+
+  return inst_list
+
+
+def get_updated_wire_names(ctrl_instance: InstanceList) -> List[str]:
+  """ return the modified wires when updating ctrl instances"""
```

Comment:
This is not following the Python doc string convention. See https://peps.python.org/pep-0257/.